pymongo/message.py

```diff
@@ -878,6 +923,305 @@ def _do_batched_op_msg(
+        buf.write(_dict_to_bson(op, False, opts))
+    resume_location = buf.tell()
+    # Write type 1 section size
+    length = buf.tell()
+    buf.seek(size_location)
+    buf.write(_pack_int(length - size_location))
+    buf.seek(resume_location)
+
+    # Type 1 Section for nsInfo
+    buf.write(b"\x01")
+    size_location = buf.tell()
+    # Save space for size
+    buf.write(b"\x00\x00\x00\x00")
+    buf.write(b"nsInfo\x00")
+    # Write all the nsInfo documents
+    for ns in to_send_ns:
+        buf.write(_dict_to_bson(ns, False, opts))
+    # Write type 1 section size
+    length = buf.tell()
+    buf.seek(size_location)
+    buf.write(_pack_int(length - size_location))
+
+    return length
+
+
+def _client_batched_op_msg_impl(
+    command: Mapping[str, Any],
+    operations: list[tuple[str, Mapping[str, Any]]],
+    ack: bool,
+    opts: CodecOptions,
+    ctx: _ClientBulkWriteContext,
+    buf: _BytesIO,
+) -> tuple[list[Mapping[str, Any]], list[Mapping[str, Any]], int]:
+    """Create a batched OP_MSG write for client-level bulk write."""
+
+    def _check_doc_size_limits(
+        op_type: str,
+        document: Mapping[str, Any],
+        limit: int,
+    ) -> int:
+        doc_size = len(_dict_to_bson(document, False, opts))
+        if doc_size > limit:
+            _raise_document_too_large(op_type, doc_size, limit)
+        return doc_size
+
+    max_bson_size = ctx.max_bson_size
+    max_write_batch_size = ctx.max_write_batch_size
+    max_message_size = ctx.max_message_size
+
+    command_doc = _dict_to_bson(command, False, opts)
+    # Don't include command-agnostic fields in document size calculations.
+    abridged_keys = ["bulkWrite", "errorsOnly", "ordered"]
+    if command.get("bypassDocumentValidation"):
+        abridged_keys.append("bypassDocumentValidation")
+    if command.get("comment"):
+        abridged_keys.append("comment")
+    if command.get("let"):
+        abridged_keys.append("let")
+    command_abridged = {key: command[key] for key in abridged_keys}
+    command_len_abridged = len(_dict_to_bson(command_abridged, False, opts))
+
+    # When OP_MSG is used unacknowledged we have to check command
+    # document size client-side or applications won't be notified.
+    if not ack:
+        _check_doc_size_limits("bulkWrite", command_abridged, max_bson_size + _COMMAND_OVERHEAD)
+
+    # Maximum combined size of the ops and nsInfo document sequences.
+    max_doc_sequences_bytes = max_message_size - (_OP_MSG_OVERHEAD + command_len_abridged)
+
+    ns_info = {}
+    to_send_ops: list[Mapping[str, Any]] = []
+    to_send_ns: list[Mapping[str, int]] = []
+    total_ops_length = 0
+    total_ns_length = 0
+    idx = 0
+
+    for real_op_type, op_doc in operations:
+        op_type = real_op_type
+        # Check insert/replace document size if unacknowledged.
+        if real_op_type == "insert":
+            if not ack:
+                _check_doc_size_limits(real_op_type, op_doc["document"], max_bson_size)
```

Comment:
We should only need to bson encode each operation once, but _check_doc_size_limits causes the op to be encoded a second time. Could you open a ticket to improve that? We don't need to do it in this PR since it's just a perf issue.

Reply:
Yep will do!